great_expectations/datasource/fluent/invalid_datasource.py

```diff
@@ -0,0 +1,176 @@
+        self, partitioner: Partitioner | None = None
+    ) -> NoReturn:
+        self._raise_type_error()
+
+
+class InvalidAssetTypeLookup(TypeLookup):
+    """A TypeLookup that always returns InvalidAsset for any type."""
+
+    @overload
+    def __getitem__(self, key: str) -> Type:
+        ...
+
+    @overload
+    def __getitem__(self, key: Type) -> str:
+        ...
+
+    @override
+    def __getitem__(self, key: ValidTypes) -> ValidTypes:
+        if isinstance(key, str):
+            return InvalidAsset
+        # if a type is passed, normally we would return the type name but that doesn't make sense here
+        # for an InvalidAsset
+        raise NotImplementedError(
+            f"Looking up the `type` name for {InvalidAsset.__name__} is not supported"
+        )
+
+
+class InvalidDatasource(Datasource):
+    """
+    A Datasource that is invalid.
+
+    This is used to represent a Datasource that is invalid and cannot be used.
+
+    This class should override all methods that would commonly be called when a user intends to use the Datasource.
+    The overridden methods should indicate to the user that the Datasource configuration is invalid and provide details about
+    why it was considered invalid.
+
+    Any errors raised should raise `from self.config_error`.
+    """
+
+    # class var definitions
+    asset_types: ClassVar[List[Type[DataAsset]]] = [InvalidAsset]
+    _type_lookup: ClassVar[TypeLookup] = InvalidAssetTypeLookup()
+
+    type: str = "invalid"
+    config_error: Union[pydantic.ValidationError, LookupError] = Field(
+        ..., description="The error that caused the Datasource to be invalid."
+    )
+    assets: List[InvalidAsset] = []
+
+    class Config:
+        extra = "ignore"
+        arbitrary_types_allowed = True
+        json_encoders = {
+            pydantic.ValidationError: lambda v: v.errors(),
+            LookupError: lambda v: repr(v),
+        }
+
+    @override
+    def test_connection(self, test_assets: bool = True) -> None:
+        raise TestConnectionError(
+            "This Datasource configuration is invalid and cannot be used. Please fix the error and try again"
+        ) from self.config_error
+
+    @override
+    def get_asset(self, asset_name: str) -> InvalidAsset:
+        """
+        Always raise a warning and return an InvalidAsset.
+        Don't raise an error because the users may want to inspect the asset config.
+        """
+        warnings.warn(
+            f"The {self.name} Datasource configuration is invalid and cannot be used. Please fix the error and try again",
+            GxInvalidDatasourceWarning,
+        )
+        return super().get_asset(asset_name)
+
+    @override
+    def get_batch_list_from_batch_request(
+        self, batch_request: BatchRequest
+    ) -> NoReturn:
+        raise TypeError(f"{self.name} Datasource is invalid") from self.config_error
+
+    @override
+    def add_batch_config(self, batch_config: BatchConfig) -> NoReturn:
```

Comment:
You could implement `__getattr__` to catch any other calls that are made on the datasource.

Reply:
Yeah I'll do this. We did something similar for block-style datasources when users tries to call a Fluent method.